### PR TITLE
[SPARK-38762][SQL] Provide query context in Decimal overflow errors

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Provide query context in Decimal overflow errors:
* explicit casting other data types as decimal
* implicit casting in decimal operations, including add/subtract/multiply/divide/reminder/pmod

### Why are the changes needed?

Provide SQL query context of runtime errors to users, so that they can understand it better.

### Does this PR introduce _any_ user-facing change?

Yes, improve the runtime error message of decimal overflow

### How was this patch tested?

Existing UT
Also test `sql/core/src/test/resources/sql-tests/inputs/cast.sql` under ANSI mode.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala [hunk 1]
```diff
@@ -168,23 +176,31 @@ case class CheckOverflowInSum(
...
+        origin.context)
     }
   }
 
   override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val childGen = child.genCode(ctx)
+    val errorContextCode = if (nullOnOverflow) {
+      ctx.addReferenceObj("errCtx", origin.context)
+    } else {
+      "null"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala:196]: shall we pass empty string as that's the default parameter value?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 2]
```diff
@@ -335,6 +335,8 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
   // The brackets that are used in casting structs and maps to strings
   private val (leftBracket, rightBracket) = if (legacyCastToStr) ("[", "]") else ("{", "}")
 
+  private var codegenContext: Option[CodegenContext] = None
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:338]: how hard is it to propagate the codegen context parameter through the function calling chain?

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:338]: It requires adding a new parameter to the following functions...I prefer the current way
```
    case DateType => castToDateCode(from, ctx)
    case decimal: DecimalType => castToDecimalCode(from, decimal, ctx)
    case TimestampType => castToTimestampCode(from, ctx)
    case TimestampNTZType => castToTimestampNTZCode(from, ctx)
    case CalendarIntervalType => castToIntervalCode(from)
    case it: DayTimeIntervalType => castToDayTimeIntervalCode(from, it)
    case it: YearMonthIntervalType => castToYearMonthIntervalCode(from, it)
    case BooleanType => castToBooleanCode(from)
    case ByteType => castToByteCode(from, ctx)
    case ShortType => castToShortCode(from, ctx)
    case IntegerType => castToIntCode(from, ctx)
    case FloatType => castToFloatCode(from, ctx)
    case LongType => castToLongCode(from, ctx)
    case DoubleType => castToDoubleCode(from, ctx)

    case array: ArrayType =>
      castArrayCode(from.asInstanceOf[ArrayType].elementType, array.elementType, ctx)
    case map: MapType => castMapCode(from.asInstanceOf[MapType], map, ctx)
    case struct: StructType => castStructCode(from.asInstanceOf[StructType], struct, ctx)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/decimalExpressions.scala]: sure, updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala [hunk 3]
```diff
@@ -1530,7 +1530,7 @@ abstract class RoundBase(child: Expression, scale: Expression,
         if (_scale >= 0) {
           s"""
             ${ev.value} = ${ce.value}.toPrecision(${ce.value}.precision(), $s,
-            Decimal.$modeStr(), true);
+            Decimal.$modeStr(), true, null);
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:1533]: shall we pass empty string instead of null here.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:1533]: Sure, updated

**Comment by gengliangwang** [comment]
Merging to master/3.3

**Pull request closed without merging**